brainiak/matnormal/utils.py

```diff
@@ -0,0 +1,73 @@
+import functools  # https://danijar.com/structuring-your-tensorflow-models/
+import tensorflow as tf
+from scipy.stats import norm
+from numpy.linalg import cholesky
+
+
+def rmn(rowcov, colcov):
+    # generate random draws from a zero-mean matrix-normal distribution
+    Z = norm.rvs(norm.rvs(size=(rowcov.shape[0], colcov.shape[0])))
+    return(cholesky(rowcov).dot(Z).dot(cholesky(colcov)))
+
+
+def doublewrap(function):
+    """
+    A decorator decorator, allowing to use the decorator to be used without
```

Comment:
See https://web.archive.org/web/20181116012513/https://danijar.com/structuring-your-tensorflow-models/. This is basically some fancy python that allows the function to create the TF graph node when it is first called and then return a handle to the same node in future times rather than create a new node every time things like the logp node gets called. I think the right thing to do here is to pull the danijar code into a separate file and make it clear this is not written by me.

Reply:
this is completely gone now, resolving.